test/py/test_migration.py

```diff
@@ -276,181 +314,193 @@ def test_migration_get_state():
 
 
-def test_handle_mig_data_write_invalid_state():
+def test_handle_mig_data_write_too_long():
+    """
+    When we set up the tests at the top of this file we specify that the max
+    data transfer size is 4 bytes. Here we test to check that a transfer of 8
+    bytes fails.
+    """
+
+    data = bytes([1, 2, 3, 4, 5, 6, 7, 8])
+
     payload = vfio_user_mig_data(
-        argsz=len(vfio_user_mig_data()) + 4,
-        size=4
+        argsz=len(vfio_user_mig_data()) + len(data),
+        size=len(data)
     )
 
-    data = bytes([1, 2, 3, 4])
-
-    transition_to_state(VFIO_USER_DEVICE_STATE_RUNNING)
+    transition_to_state(ctx, sock, VFIO_USER_DEVICE_STATE_RESUMING)
     msg(ctx, sock, VFIO_USER_MIG_DATA_WRITE, bytes(payload) + data,
         expect=errno.EINVAL)
 
 
 def test_handle_mig_data_write_failed_callback():
-    global fail_callbacks
+    transition_to_state(ctx, sock, VFIO_USER_DEVICE_STATE_RESUMING)
 
-    transition_to_state(VFIO_USER_DEVICE_STATE_RESUMING)
+    data = bytes([1, 2, 3, 4])
 
-    fail_callbacks = True
+    setup_fail_callbacks()
 
     payload = vfio_user_mig_data(
-        argsz=len(vfio_user_mig_data()) + 4,
-        size=4
+        argsz=len(vfio_user_mig_data()) + len(data),
+        size=len(data)
     )
 
-    data = bytes([1, 2, 3, 4])
-
     msg(ctx, sock, VFIO_USER_MIG_DATA_WRITE, bytes(payload) + data,
         expect=errno.EINVAL)
+    assert c.get_errno() == errno.EINVAL
 
-    fail_callbacks = False
+    teardown_fail_callbacks()
 
 
 def test_handle_mig_data_write_short_write():
-    payload = struct.pack("I", 8)
+    data = bytes([1, 2, 3, 4])
+
+    payload = vfio_user_mig_data(
+        argsz=len(vfio_user_mig_data()) + len(data),
+        size=len(data)
+    )
 
     msg(ctx, sock, VFIO_USER_MIG_DATA_WRITE, payload, expect=errno.EINVAL)
 
 
-def test_device_feature_migration():
+def test_device_feature_migration_get():
     payload = vfio_user_device_feature(
-        argsz=len(vfio_user_device_feature()) + 8,
+        argsz=len(vfio_user_device_feature()) +
+              len(vfio_user_device_feature_migration()),
         flags=VFIO_DEVICE_FEATURE_GET | VFIO_DEVICE_FEATURE_MIGRATION
     )
 
     result = msg(ctx, sock, VFIO_USER_DEVICE_FEATURE, payload)
     _, result = vfio_user_device_feature.pop_from_buffer(result)
-    flags, = struct.unpack("Q", result)
+    flags, _ = vfio_user_device_feature_migration.pop_from_buffer(result)
+    flags = flags.flags
 
     assert flags == VFIO_MIGRATION_STOP_COPY | VFIO_MIGRATION_PRE_COPY
 
 
 def test_device_feature_short_write():
-    payload = struct.pack("I", 8)
+    payload = vfio_user_device_feature(
+        argsz=len(vfio_user_device_feature()) +
+              len(vfio_user_device_feature_migration()),
+        flags=VFIO_DEVICE_FEATURE_GET | VFIO_DEVICE_FEATURE_MIGRATION
+    )
 
-    msg(ctx, sock, VFIO_USER_DEVICE_FEATURE, payload, expect=errno.EINVAL)
+    payload = bytes(payload)
+    assert len(payload) == 8
```

Comment:
8?

Reply:
This was to make it clear that the payload has length 8 and we're sending 7 bytes of it, I've changed this to now explicitly send one byte less than the payload length which I should've done in the first place!